### PR TITLE
A fix for issues in getdents64 and getcwd related to non-ascii file names

### PR DESCRIPTION
**Pull request by ImplOfAnImpl**
Hi, this PR fixes a couple of issues:
1) if the current directory name contains non-ascii characters, getcwd returns garbage.
2) similarly, getdents returns garbage for files with non-ascii names.

The latter problem impacts boost::filesystem, e.g. boost::filesystem::remove_all fails to remove a directory with non-ascii-named files.

Test suites run: main, other (got some failures in 'other', but they were the same without my changes too).

### CODE REVIEW

File: tests/fs/test_getcwd_with_non_ascii_name.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+{
+    const auto* const dirName = u8"абвг";
+
+    auto ret = mkdir(dirName, 0777);
+    assert(ret == 0);
+    ret = chdir(dirName);
+    assert(ret == 0);
+
+    char buf[255];
+    auto ret2 = getcwd(buf, sizeof(buf));
```
Comment (juj) [inline tests/fs/test_getcwd_with_non_ascii_name.cpp]: Could you update this test to try out a longer path name, and also test the negative case of truncation if the path exceeds `sizeof(buf)`, and verify that this does not overflow the destination space?

File: src/library_syscall.js [hunk 2]
```diff
@@ -996,7 +997,8 @@ var SyscallsLibrary = {
       var id;
       var type;
       var name = stream.getdents.pop();
-      assert(name.length < 256); // limit of dirent struct
+      var nameLengthInBytes = lengthBytesUTF8(name);
+      assert(nameLengthInBytes < 256); // limit of dirent struct
```
Comment (juj) [inline src/library_syscall.js]: Hmm, this `assert` actually seems wrong. We can just remove it, and depend on the line `stringToUTF8(name, dirp + pos + {{{ C_STRUCTS.dirent.d_name }}}, 256);` to do its proper job with truncation.

**Comment by juj** [comment]
Thanks, looks like good fixes, though let's add testing for the truncation cases as well, which would have previously failed/produced bad results.

File: tests/test_core.py [hunk 3]
```diff
@@ -4122,17 +4122,17 @@ def test_getdents64(self):
     src = open(path_from_root('tests', 'fs', 'test_getdents64.cpp'), 'r').read()
     self.do_run(src, '..')
 
+  def test_getdents64_special_cases(self):
+    Building.COMPILER_TEST_OPTS += ['-std=c++11']
+    src = path_from_root('tests', 'fs', 'test_getdents64_special_cases.cpp')
+    out = path_from_root('tests', 'fs', 'test_getdents64_special_cases.out')
+    self.do_run_from_file(src, out, assert_identical=True)
```
Comment (ImplOfAnImpl) [inline tests/test_core.py]: FYI: assert_identical is needed here because otherwise the line "doTest("test_dir3", std::string(254, '1') + u8"今日は");" would succeeded even if I deliberately broke stringToUTF8Array.

File: tests/runner.py [hunk 4]
```diff
@@ -663,9 +664,13 @@ def do_run(self, src, expected_output, args=[], output_nicerizer=None,
...
       js_output = self.run_generated_code(engine, filename + '.o.js', args, output_nicerizer=output_nicerizer, assert_returncode=assert_returncode)
+      js_output = js_output.replace('\r\n', '\n')
       try:
-        self.assertContained(expected_output, js_output.replace('\r\n', '\n'))
-        self.assertNotContained('ERROR', js_output)
+        if assert_identical:
+          self.assertIdentical(expected_output, js_output)
+        else:
+          self.assertContained(expected_output, js_output)
+          self.assertNotContained('ERROR', js_output)
```
Comment (juj) [inline tests/runner.py]: Hmm, how come updating the test required changing the harness?

File: tests/runner.py [hunk 5]
```diff
@@ -663,9 +664,13 @@ def do_run(self, src, expected_output, args=[], output_nicerizer=None,
     for engine in js_engines:
       #print 'test in', engine
       js_output = self.run_generated_code(engine, filename + '.o.js', args, output_nicerizer=output_nicerizer, assert_returncode=assert_returncode)
+      js_output = js_output.replace('\r\n', '\n')
```
Comment (juj) [inline tests/runner.py]: We should have many multiline tests before, is there something special that required changing this flow?

File: src/library_syscall.js [hunk 6]
```diff
@@ -871,8 +871,9 @@ var SyscallsLibrary = {
     var buf = SYSCALLS.get(), size = SYSCALLS.get();
     if (size === 0) return -ERRNO_CODES.EINVAL;
     var cwd = FS.cwd();
-    if (size < cwd.length + 1) return -ERRNO_CODES.ERANGE;
-    writeAsciiToMemory(cwd, buf);
+    var cwdLengthInBytes = lengthBytesUTF8(cwd);
+    if (size < cwdLengthInBytes + 1) return -ERRNO_CODES.ERANGE;
+    stringToUTF8(cwd, buf, cwdLengthInBytes + 1);
```
Comment (juj) [inline src/library_syscall.js]: I think there is an issue with this line. Shouldn't it be `stringToUTF8(cwd, buf, size);`? I wonder why this was not caught in the test?

File: tests/runner.py (hunk 4 above)
Comment (ImplOfAnImpl) [inline tests/runner.py]: If the last test in test_getdents64_special_cases fails because of a partial character being written at the end of the string (e.g. if stringToUTF8 is broken), its output will contain the expected output plus a garbage character and assertContained will succeed. For this not to happen I could:
1) Reorder the calls to doTest. This is too non-obvious and unreliable IMO,
2) Print something at the end of the test (e.g. "done") and add the same string into the expected output. This is how I did it initially, but it seemed rather unreliable too - the essential part here is the "done" string in the expected output and I can't put a comment there to mention its importance.
So, introducing the "assert_identical" mode looked the best to me.

File: tests/runner.py (hunk 5 above)
Comment (ImplOfAnImpl) [inline tests/runner.py]: `js_output.replace` was already called in `self.assertContained(expected_output, js_output.replace('\r\n', '\n'))`, I just moved it out of the condition to avoid minor code duplication.

The only difference with the old logic is that the search for 'ERROR' is now performed in the 'replaced' js_output, while initially it was done in the original js_output.

File: src/library_syscall.js (hunk 6 above)
Comment (ImplOfAnImpl) [inline src/library_syscall.js]: Well, the condition above guarantees that `size >= cwdLengthInBytes + 1`, so there is no error here. But I can change it to `size` if you think that the code will be more clear.

**Comment by ImplOfAnImpl** [comment]
So, what's the verdict?

File: src/library_syscall.js (hunk 6 above)
Comment (juj) [reply src/library_syscall.js]: Ah, you are definitely right. Though it would still be good to do this change, to explicitly use the size that the buffer actually is to be clear.

File: tests/runner.py (hunk 4 above)
Comment (juj) [reply tests/runner.py]: We can assume `stringToUTF8` to never output partial characters at all, but always properly truncate to full characters. That should be the case at https://github.com/kripken/emscripten/blob/master/src/preamble.js#L623, so we do not need to complicate for testing against a broken `stringToUTF8` here. Or are you saying you found a problem with `stringToUTF8` and had to work around that with the changes here?

File: tests/test_core.py (hunk 3 above)
Comment (juj) [reply tests/test_core.py]: Reading the changes to the test harness, introducing `assert_identical=True` looks ok to me, though @kripken, would be good for you to also take a glance for that part?

**Comment by juj** [comment]
Sorry for the delay here. The PR looks good, if you can do the size change, I'd be alright to merge. Though let's wait a moment for @kripken to comment as well.

File: tests/runner.py (hunk 4 above)
Comment (ImplOfAnImpl) [reply tests/runner.py]: No, there were no problems with stringToUTF8.

OK, I'll remove the test for a bad truncation. Should I remove the "assert_identical" mode too?
My opinion - it's still useful, since it allows the test implementer to better express their intention.

Comment (juj) [reply tests/runner.py]: I think the assert_identical mode looks useful to have, but let's wait for @kripken for his thoughts as well.

Comment (kripken) [reply tests/runner.py]: Yeah, seems useful to me too.

**Comment by kripken** [comment]
No concerns from me.

**Comment by juj** [comment]
Perfec,t thanks!
